File: python/mirage/core/gsheets/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import posixpath

from mirage.accessor.gsheets import GSheetsAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.gsheets._client import (SHEETS_API_BASE, TokenManager,
                                         google_get)
from mirage.core.gsheets.readdir import readdir
from mirage.types import PathSpec


async def read_spreadsheet(token_manager: TokenManager,
                           spreadsheet_id: str) -> bytes:
    """Fetch full spreadsheet JSON.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.

    Returns:
        bytes: JSON response as bytes.
    """
    url = f"{SHEETS_API_BASE}/spreadsheets/{spreadsheet_id}"
    data = await google_get(token_manager, url)
    return json.dumps(data, ensure_ascii=False, separators=(",", ":")).encode()


async def read_values(token_manager: TokenManager, spreadsheet_id: str,
                      range_: str) -> bytes:
    """Read cell values via Values API. Returns JSON array.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.
        range_ (str): A1 notation range.

    Returns:
        bytes: JSON response as bytes.
    """
    url = f"{SHEETS_API_BASE}/spreadsheets/{spreadsheet_id}/values/{range_}"
    data = await google_get(token_manager, url)
    return json.dumps(data, ensure_ascii=False, separators=(",", ":")).encode()


async def fetch_sheet_names(
    token_manager: TokenManager,
    spreadsheet_id: str,
) -> list[str]:
    """Fetch sheet tab names for a spreadsheet.

    Args:
        token_manager (TokenManager): manages OAuth2 tokens.
        spreadsheet_id (str): Google Sheets spreadsheet ID.

    Returns:
        list[str]: list of sheet tab names.
    """
    fields = "sheets.properties.title"
    url = f"{SHEETS_API_BASE}/spreadsheets/{spreadsheet_id}?fields={fields}"
    data = await google_get(token_manager, url)
    return [s["properties"]["title"] for s in data.get("sheets", [])]


async def read(
    accessor: GSheetsAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    if index is None:
        raise FileNotFoundError(path)
    virtual_key = prefix + "/" + key if prefix else "/" + key
    result = await index.get(virtual_key)
    if result.entry is None:
        parent_key = posixpath.dirname(virtual_key) or "/"
        if parent_key != virtual_key:
            parent_path = PathSpec.from_str_path(parent_key, prefix=prefix)
            try:
                await readdir(accessor, parent_path, index)
                result = await index.get(virtual_key)
            except Exception:
                pass
        if result.entry is None:
            raise FileNotFoundError(path)
    return await read_spreadsheet(accessor.token_manager, result.entry.id)

File: typescript/packages/core/src/core/gsheets/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GSheetsAccessor } from '../../accessor/gsheets.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { SHEETS_API_BASE, type TokenManager, googleGet } from '../google/_client.ts'
import { readdir } from './readdir.ts'
import { rstripSlash, stripSlash } from '../../util/slash.ts'

const ENC = new TextEncoder()

function enoent(p: string): Error {
  const e = new Error(`ENOENT: ${p}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

export async function readSpreadsheet(
  tm: TokenManager,
  spreadsheetId: string,
): Promise<Uint8Array> {
  const url = `${SHEETS_API_BASE}/spreadsheets/${spreadsheetId}`
  const data = await googleGet(tm, url)
  return ENC.encode(JSON.stringify(data))
}

export async function readValues(
  tm: TokenManager,
  spreadsheetId: string,
  range: string,
): Promise<Uint8Array> {
  const url = `${SHEETS_API_BASE}/spreadsheets/${spreadsheetId}/values/${range}`
  const data = await googleGet(tm, url)
  return ENC.encode(JSON.stringify(data))
}

export async function fetchSheetNames(tm: TokenManager, spreadsheetId: string): Promise<string[]> {
  const fields = 'sheets.properties.title'
  const url = `${SHEETS_API_BASE}/spreadsheets/${spreadsheetId}?fields=${fields}`
  const data = (await googleGet(tm, url)) as {
    sheets?: { properties?: { title?: string } }[]
  }
  const out: string[] = []
  for (const s of data.sheets ?? []) {
    if (s.properties?.title !== undefined) out.push(s.properties.title)
  }
  return out
}

export async function read(
  accessor: GSheetsAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (index === undefined) throw enoent(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`
  let result = await index.get(virtualKey)
  if (result.entry === undefined || result.entry === null) {
    const parentKey = rstripSlash(virtualKey).replace(/\/[^/]+$/, '') || '/'
    if (parentKey !== virtualKey) {
      const parentPath = PathSpec.fromStrPath(parentKey, prefix)
      try {
        await readdir(accessor, parentPath, index)
        result = await index.get(virtualKey)
      } catch {
        // parent refresh failed; fall through to ENOENT
      }
    }
    if (result.entry === undefined || result.entry === null) throw enoent(path.original)
  }
  return readSpreadsheet(accessor.tokenManager, result.entry.id)
}

export async function* stream(
  accessor: GSheetsAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  yield await read(accessor, path, index)
}
